R/utilities-sensitivity-analysis.R

```diff
@@ -1,342 +1,343 @@
+  sortVec <- sort(seqVec) # Sort seqVec to obtain an concatenated array of repeated integers, with the repeated integers ranging from from 1 to numberOfCores.  These are the core numbers to which each parameter will be assigned.
+  listSplitParameters <- split(x = parametersToPerturb, sortVec) # Split the parameters of the model according to sortVec
+  # listSplitParameters <- split(x = parametersToPerturb, sort(1+( (1:length(totalNumberParameters)) %% numberOfCores))  )
+  Rmpi::mpi.spawn.Rslaves(nslaves = numberOfCores)
+  Rmpi::mpi.bcast.cmd(library("ospsuite"))
+  Rmpi::mpi.bcast.cmd(library("ospsuite.reportingengine"))
+  Rmpi::mpi.bcast.Robj2slave(obj = simFilePath)
+  Rmpi::mpi.bcast.Robj2slave(obj = listSplitParameters)
+  Rmpi::mpi.bcast.Robj2slave(obj = resultsFileFolder)
+  Rmpi::mpi.bcast.Robj2slave(obj = individualParameters)
+
+  # Generate a listcontaining names of SA CSV result files that will be output by each core
+  allResultsFileNames <- generateResultFileNames(numberOfCores = numberOfCores, folderName = resultsFileFolder, fileName = resultsFileName)
+
+
+  Rmpi::mpi.bcast.Robj2slave(obj = allResultsFileNames)
+
+  # Load simulation on each core
+  Rmpi::mpi.bcast.cmd(sim <- loadSimulation(simFilePath))
+
+  # Update simulation with individual parameters
+  Rmpi::mpi.bcast.cmd(updateSimulationIndividualParameters(simulation = sim, individualParameters))
+
+
+
+  Rmpi::mpi.remote.exec(analyzeCoreSensitivity(
+    simulation = sim,
+    parametersToPerturb = listSplitParameters[[mpi.comm.rank()]],
+    totalSensitivityThreshold = 1,
+    resultsFilePath = allResultsFileNames[mpi.comm.rank()],
+    numberOfCoresToUse = 1 # Number of local cores, set to 1 when parallelizing.
+  ))
+  Rmpi::mpi.close.Rslaves()
+
+  allSAResults <- importSensitivityAnalysisResultsFromCSV(simulation = loadSimulation(simFilePath), filePaths = allResultsFileNames)
+  combinedFilePath <- file.path(resultsFileFolder, paste0(resultsFileName, ".csv"))
+  exportSensitivityAnalysisResultsToCSV(results = allSAResults, filePath = combinedFilePath)
+  file.remove(allResultsFileNames)
+  return(combinedFilePath)
+}
+
+#' @title analyzeCoreSensitivity
+#' @description Run a sensitivity analysis for a single individual, perturbing only the set of parameters parametersToPerturb
+#' @param simulation simulation class object
+#' @param parametersToPerturb paths of parameters to be analyzed
+#' @param totalSensitivityThreshold numeric value between 0 and 1.
+#' Close to 0, only the most sensitive output paths are returned.
+#' Close to 1, almost all the output paths are returned.
+#' @param resultsFilePath Path to file storing results of sensitivity analysis
+#' @param numberOfCoresToUse Number of cores to use on local node.  This parameter should be should be set to 1 when parallelizing over many nodes.
+#' @return Save sensitivity analysis results as CSV in path given by resultsFilePath.
+#' @import ospsuite
+#' @export
```

Comment:
Do we need to export it?

Reply:
Yes this one should be exported for access by cores when parallelizing